### ISSUE TITLE
Rule proposal: Consistent function scoping

### ISSUE THREAD

**Issue by futpib** [sindresorhus/eslint-plugin-unicorn#230]
[<img alt="Issuehunt badges" src="https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/230/badge.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/230)

A function definition should be placed as close to the top level scope as possible without breaking it's captured values.

## Fail

```js
export function create(context) {
  // does not capture anything from the scope, can be moved to the outer scope
  function isSpecial(node) {
    return node.type === 'Special';
  }

  return isSpecial;
}
```

## Pass

```js
function isSpecial(node) {
  return node.type === 'Special';
}

export function create(context) {
  return isSpecial;
}
```

```js
export function create(context) {
  function isSpecial(node) {
    return node.type === 'Special' && context.isSpecial(node);
  }

  return isSpecial;
}
```

> There is a $80.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/230).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

**Comment by sindresorhus** [comment]
Sounds good. I often have to manually comment about this in PR reviews.

**Comment by sindresorhus** [comment]
Slightly off-topic, but I would like the inverse for variables. That they are placed in the most inner scope as possible. What do you think?

**Comment by futpib** [comment]
This might actually be perfectly on-topic, because of the interaction between the two rules.

In this example, should `context` be moved in, or `isSpecial` be moved out? (I have no idea yet)

```js
const context = undefined;

export function create() {
  function isSpecial(node) {
    return node.type === 'Special' && context.isSpecial(node);
  }

  return isSpecial;
}
```

**Comment by futpib** [comment]
On a second thought moving variables in may be a bad idea because it may negatively affect performance.

**Comment by sindresorhus** [comment]
Yeah, let's keep this focused on only functions for now. And yes it should be moved out.

**Comment by IssuehuntBot** [comment]
@issuehunt has funded $80.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/230) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/230).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by MrHen** [comment]
I have a prototype for this working; hope to have a PR out in the next couple days.

**Comment by golopot** [comment]
Alternative names:
* function-scoping
* function-lifting
* function-hoisting
